File: julia/test/semantic_index_static_remaining_test.jl
const SEMANTIC_STATIC_REMAINING_ROOT = normpath(
    joinpath(@__DIR__, "..", "..", "capability_conformance", "semantic_introspection"),
)

const SEMANTIC_STATIC_REMAINING_MODEL = JSON3.read(
    read(
        joinpath(
            SEMANTIC_STATIC_REMAINING_ROOT,
            "..",
            "semantic_introspection_model.json",
        ),
        String,
    ),
    Dict{String,Any},
)

function _semantic_static_remaining_index(
    fixture::String,
    ceiling::SemanticSourceDetail,
)
    return semantic_index(
        read(joinpath(SEMANTIC_STATIC_REMAINING_ROOT, "$fixture.spec"), String);
        logical_name = "$fixture.spec",
        source_detail_ceiling = ceiling,
    )
end

function _semantic_static_remaining_expected(snapshot_id::String)
    snapshot = only(
        candidate for candidate in SEMANTIC_STATIC_REMAINING_MODEL["snapshots"] if
        candidate["id"] == snapshot_id
    )
    result = JSON3.read(JSON3.write(snapshot), Dict{String,Any})
    delete!(result, "id")
    delete!(result, "fixture")
    return result
end

function _semantic_static_remaining_materialize(projection::Dict{String,Any})
    result = JSON3.read(JSON3.write(projection), Dict{String,Any})
    source_refs = pop!(result, "source_refs")
    for group in ("records", "relations")
        for item in result[group]
            source = item["source"]
            if source isa AbstractString
                item["source"] = JSON3.read(
                    JSON3.write(source_refs[source]),
                    Dict{String,Any},
                )
            end
        end
    end
    return result
end

function _semantic_static_remaining_runtime_expected()
    wanted = _semantic_static_remaining_expected("runtime")
    wanted["records"] = [
        record for record in wanted["records"] if
        !(record["kind"] in ("execution", "event"))
    ]
    retained = Set(record["id"] for record in wanted["records"])
    wanted["relations"] = [
        relation for relation in wanted["relations"] if
        relation["from_id"] in retained && relation["to_id"] in retained
    ]
    wanted["snapshot"]["has_execution"] = false
    return wanted
end

function _semantic_static_remaining_record(projection, record_id)
    return only(record for record in projection["records"] if record["id"] == record_id)
end

function _semantic_static_remaining_plain(value)
    forbidden_keys = Set([
        "source_text",
        "source_bytes",
        "ast",
        "action_ir",
        "descriptor",
        "compiled_regex",
        "executor",
        "trace",
        "diagnostic_sink",
        "runtime_observer",
    ])
    if value isa AbstractDict
        return all(key isa AbstractString && !(key in forbidden_keys) for key in keys(value)) &&
               all(_semantic_static_remaining_plain, values(value))
    elseif value isa AbstractVector
        return all(_semantic_static_remaining_plain, value)
    end
    return value === nothing || value isa AbstractString || value isa Number || value isa Bool
end

@testset "Semantic index remaining private static targets and isolation" begin
    @testset "Unicode privacy ceilings" begin
        for (snapshot_id, ceiling, expected_ceiling, digest_available) in (
            ("privacy", SemanticSourceTextDetail, "text", true),
            ("privacy_limited", SemanticSourceIdentityDetail, "identity", false),
        )
            index = _semantic_static_remaining_index("privacy", ceiling)
            raw = LinkedSpecJulia._semantic_static_projection_for_testing(index)
            actual = _semantic_static_remaining_materialize(raw)
            wanted = _semantic_static_remaining_materialize(
                _semantic_static_remaining_expected(snapshot_id),
            )
            @test actual == wanted
            @test length(actual["records"]) == 4
            @test length(actual["relations"]) == 3
            @test actual["snapshot"]["source_detail_ceiling"] == expected_ceiling
            @test actual["snapshot"]["content_digest_available"] == digest_available
            @test length(raw["source_refs"]) == 2

            rule = _semantic_static_remaining_record(actual, "rule:T%C3%B6p")
            slot = _semantic_static_remaining_record(actual, "regex:rule:T%C3%B6p:0")
            @test rule["name"] == "Töp"
            @test rule["source"]["span"]["start_byte"] == 0
            @test rule["source"]["span"]["end_byte"] == 6
            @test rule["source"]["excerpt"] == "Töp::"
            @test slot["facts"]["pattern"] == "é"
            @test slot["source"]["span"]["start_column"] == 2
            @test slot["source"]["span"]["end_column"] == 5
        end
    end

    @testset "runtime static absence" begin
        index = _semantic_static_remaining_index("runtime", SemanticSourceTextDetail)
        actual = _semantic_static_remaining_materialize(
            LinkedSpecJulia._semantic_static_projection_for_testing(index),
        )
        wanted = _semantic_static_remaining_materialize(
            _semantic_static_remaining_runtime_expected(),
        )
        @test actual == wanted
        @test length(actual["records"]) == 7
        @test length(actual["relations"]) == 8
        @test actual["snapshot"]["has_execution"] == false
        @test !any(record["kind"] in ("execution", "event") for record in actual["records"])
        @test !any(relation["kind"] == "observed_as" for relation in actual["relations"])
        @test !any(relation["kind"] == "dispatches_to" for relation in actual["relations"])
        @test count(relation -> relation["kind"] == "selects_regex", actual["relations"]) == 2

        rule = _semantic_static_remaining_record(actual, "rule:Top")
        @test rule["facts"]["is_repetition"] == true
        @test rule["facts"]["rep_min"] == 2
        @test rule["facts"]["rep_max"] == 2
        @test [
            _semantic_static_remaining_record(actual, "regex:rule:Top:$index")["facts"]["pattern"] for
            index in 0:1
        ] == ["a", "b"]
    end

    @testset "failed compilation normalization" begin
        index = _semantic_static_remaining_index("failed", SemanticSourceSpanDetail)
        native = compilation_diagnostic(index)
        @test native.code == "bare_edge_target_undefined"
        @test native.stage == "normalize_edges"
        @test native.message == "bare edge in rule 'Top' targets undefined rule 'Missing'"
        @test Dict(native.fields) == Dict{String,Any}(
            "rule_label" => "Top",
            "target" => "Missing",
        )

        raw = LinkedSpecJulia._semantic_static_projection_for_testing(index)
        actual = _semantic_static_remaining_materialize(raw)
        wanted = _semantic_static_remaining_materialize(
            _semantic_static_remaining_expected("failed"),
        )
        @test actual == wanted
        @test length(actual["records"]) == 6
        @test length(actual["relations"]) == 4
        @test actual["snapshot"] == Dict{String,Any}(
            "id" => "snapshot:0",
            "state" => "failed_compilation",
            "has_execution" => false,
            "source_detail_ceiling" => "span",
            "content_digest_available" => false,
        )
        @test [record["id"] for record in actual["records"]] == [
            "spec:0",
            "source:0",
            "rule:Top",
            "diagnostic:compile:0",
            "decision:compile:rule:Top",
            "explanation:decision:compile:rule:Top:0",
        ]

        diagnostic = _semantic_static_remaining_record(actual, "diagnostic:compile:0")
        @test diagnostic["facts"]["code"] == "unknown_rule_reference"
        @test diagnostic["facts"]["stage"] == "compile"
        @test diagnostic["facts"]["fields"] == Dict{String,Any}(
            "rule_id" => "rule:Top",
            "missing_rule_id" => "rule:Missing",
        )
        @test diagnostic["source"]["excerpt"] == "Missing"
        @test diagnostic["source"]["span"]["start_byte"] == 6
        @test diagnostic["source"]["span"]["end_byte"] == 13

        rule = _semantic_static_remaining_record(actual, "rule:Top")
        @test rule["facts"]["edge_ownership"] == "action"
        @test rule["source"]["excerpt"] == "Top:"
        @test count(relation -> relation["kind"] == "diagnoses", actual["relations"]) == 1
        explained = only(
            relation for relation in actual["relations"] if
            relation["kind"] == "explained_by"
        )
        @test explained["evidence_ids"] == ["diagnostic:compile:0"]
    end

    @testset "repeated lifecycle occurrence identity" begin
        source = """Top::
 /x/
 E { return("first") }
 E { return(["second"]) }
"""
        index = semantic_index(
            source;
            logical_name = "repeated-lifecycle.spec",
            source_detail_ceiling = SemanticSourceTextDetail,
        )
        actual = _semantic_static_remaining_materialize(
            LinkedSpecJulia._semantic_static_projection_for_testing(index),
        )
        lifecycles = [
            record for record in actual["records"] if record["kind"] == "lifecycle"
        ]
        @test length(lifecycles) == 2
        @test [record["id"] for record in lifecycles] == [
            "lifecycle:rule:Top:E:0",
            "lifecycle:rule:Top:E:1",
        ]
        @test [record["order"] for record in lifecycles] == [0, 1]
        @test [record["facts"]["value_shape"]["kind"] for record in lifecycles] == [
            "string",
            "array",
        ]
        @test [record["source"]["excerpt"] for record in lifecycles] == [
            "E { return(\"first\") }",
            "E { return([\"second\"]) }",
        ]
        @test [record["source"]["span"]["start_line"] for record in lifecycles] == [3, 4]
        @test count(
            relation -> relation["kind"] == "contains" &&
                        startswith(relation["to_id"], "lifecycle:"),
            actual["relations"],
        ) == 2
    end

    @testset "clone, fallback, and host isolation" begin
        indexes = (
            _semantic_static_remaining_index("privacy", SemanticSourceTextDetail),
            _semantic_static_remaining_index("privacy", SemanticSourceIdentityDetail),
            _semantic_static_remaining_index("failed", SemanticSourceSpanDetail),
            _semantic_static_remaining_index("runtime", SemanticSourceTextDetail),
        )
        for index in indexes
            projection = LinkedSpecJulia._semantic_static_projection_for_testing(index)
            @test _semantic_static_remaining_plain(projection)
            @test JSON3.read(JSON3.write(projection), Dict{String,Any}) == projection
            @test !occursin(r"/(?:Users|home|tmp)/", JSON3.write(projection))
        end

        failed_index = indexes[3]
        first_copy = LinkedSpecJulia._semantic_static_projection_for_testing(failed_index)
        first_diagnostic = _semantic_static_remaining_record(first_copy, "diagnostic:compile:0")
        first_diagnostic["facts"]["fields"]["missing_rule_id"] = "rule:Injected"
        second_copy = LinkedSpecJulia._semantic_static_projection_for_testing(failed_index)
        second_diagnostic = _semantic_static_remaining_record(second_copy, "diagnostic:compile:0")
        @test second_diagnostic["facts"]["fields"]["missing_rule_id"] == "rule:Missing"
        @test first_copy !== second_copy

        retained = getfield(failed_index, :_static_projection)
        @test retained isa LinkedSpecJulia._SemanticStaticProjection
        @test retained.records.values isa Tuple
        @test_throws MethodError setindex!(retained.records.values, nothing, 1)

        parse_failure = semantic_index(
            "Top:::\n /x/\n";
            logical_name = "parse-failure.spec",
            source_detail_ceiling = SemanticSourceIdentityDetail,
        )
        parse_projection = LinkedSpecJulia._semantic_static_projection_for_testing(parse_failure)
        @test length(parse_projection["records"]) == 3
        @test length(parse_projection["relations"]) == 2
        @test _semantic_static_remaining_record(
            parse_projection,
            "diagnostic:compile:0",
        )["facts"]["code"] == "semantic_index_parse_failed"

        missing_entry = semantic_index(
            "Top::\n /x/\n";
            logical_name = "missing-entry.spec",
            source_detail_ceiling = SemanticSourceIdentityDetail,
            entry_rule = "Missing",
        )
        missing_projection = LinkedSpecJulia._semantic_static_projection_for_testing(missing_entry)
        @test length(missing_projection["records"]) == 4
        @test length(missing_projection["relations"]) == 2
        @test _semantic_static_remaining_record(
            missing_projection,
            "diagnostic:compile:0",
        )["facts"]["code"] == "entry_rule_not_found"

        @test !(:_semantic_static_projection_for_testing in names(LinkedSpecJulia))
        implementation = read(
            joinpath(@__DIR__, "..", "src", "semantic", "SemanticStaticProjection.jl"),
            String,
        )
        for forbidden in (
            "SpecLoader",
            "to_descriptor_json",
            "emit_julia_source",
            "execute_generated",
            "LINKEDSPEC_TRACE_LEVEL",
            "diagnostic_output_sink",
            "runtime_observer",
            "RuntimeSemanticObservation",
            "ENV[",
            "time_ns(",
            "rand(",
        )
            @test !occursin(forbidden, implementation)
        end
    end
end

File: lua/test/semantic_index_static_remaining_test.lua
-- FUTURE-PARITY-BACKLOG.10.7.3.2.1 — remaining private static targets.

local assertions = 0
local failures = {}

local function check(condition, label)
  assertions = assertions + 1
  if not condition then failures[#failures + 1] = label or "assertion failed" end
end

local function check_equal(actual, expected, label)
  check(actual == expected, (label or "values differ") ..
    ": expected " .. tostring(expected) .. ", got " .. tostring(actual))
end

local function read_file(path)
  local handle = assert(io.open(path, "rb"))
  local value = assert(handle:read("*a"))
  assert(handle:close())
  return value
end

local linkedspec = require("linkedspec")
local json = linkedspec.json
local semantic_index_module = require("linkedspec.semantic_index")

local function options(logical_name, source_detail_ceiling, entry_rule)
  return {
    logical_name = logical_name,
    source_detail_ceiling = source_detail_ceiling,
    entry_rule = entry_rule,
  }
end

local function detached(value)
  return json.decode(json.encode(value))
end

local function snapshot_by_id(model, snapshot_id)
  for _, snapshot in ipairs(model.snapshots) do
    if snapshot.id == snapshot_id then return snapshot end
  end
  error("missing semantic snapshot " .. snapshot_id, 0)
end

local function expected_snapshot(model, snapshot_id)
  local result = detached(snapshot_by_id(model, snapshot_id))
  result.id = nil
  result.fixture = nil
  return result
end

local function materialize_sources(projection)
  local result = detached(projection)
  local source_refs = result.source_refs
  result.source_refs = nil
  for _, group in ipairs({ "records", "relations" }) do
    for _, item in ipairs(result[group]) do
      if type(item.source) == "string" then item.source = detached(source_refs[item.source]) end
    end
  end
  return result
end

local function runtime_static_expected(model)
  local result = expected_snapshot(model, "runtime")
  local records = json.array()
  local retained = {}
  for _, item in ipairs(result.records) do
    if item.kind ~= "execution" and item.kind ~= "event" then
      records[#records + 1] = item
      retained[item.id] = true
    end
  end
  local relations = json.array()
  for _, item in ipairs(result.relations) do
    if retained[item.from_id] and retained[item.to_id] then relations[#relations + 1] = item end
  end
  result.records = records
  result.relations = relations
  result.snapshot.has_execution = false
  return result
end

local function record_by_id(projection, id)
  for _, item in ipairs(projection.records) do
    if item.id == id then return item end
  end
  error("missing semantic record " .. id, 0)
end

local function records_of_kind(projection, kind)
  local result = {}
  for _, item in ipairs(projection.records) do
    if item.kind == kind then result[#result + 1] = item end
  end
  return result
end

local function count_relations(projection, kind)
  local count = 0
  for _, item in ipairs(projection.relations) do
    if item.kind == kind then count = count + 1 end
  end
  return count
end

local function count_keys(value)
  local count = 0
  for _ in pairs(value) do count = count + 1 end
  return count
end

local function plain_and_host_free(value)
  if value == json.null then return true end
  local kind = json.kind(value)
  if kind == "string" or kind == "number" or kind == "boolean" then return true end
  if kind ~= "array" and kind ~= "harray" then return false end
  local forbidden = {
    source_text = true,
    source_bytes = true,
    ast = true,
    action_ir = true,
    descriptor = true,
    compiled_regex = true,
    loader = true,
    executor = true,
    trace = true,
    diagnostic_sink = true,
    runtime_observer = true,
  }
  for key, item in pairs(value) do
    if kind == "harray" and (type(key) ~= "string" or forbidden[key]) then return false end
    if not plain_and_host_free(item) then return false end
  end
  return true
end

local model = json.decode(read_file("capability_conformance/semantic_introspection_model.json"))
local fixture_root = "capability_conformance/semantic_introspection/"

for _, case in ipairs({
  { "privacy", "text", "text", true },
  { "privacy_limited", "identity", "identity", false },
}) do
  local index = linkedspec.semantic_index(
    read_file(fixture_root .. "privacy.spec"),
    options("privacy.spec", case[2])
  )
  local raw = semantic_index_module._static_projection_for_testing(index)
  local actual = materialize_sources(raw)
  local expected = materialize_sources(expected_snapshot(model, case[1]))
  check_equal(json.encode(actual), json.encode(expected), case[1] .. " deep equality")
  check_equal(#actual.records, 4, case[1] .. " record count")
  check_equal(#actual.relations, 3, case[1] .. " relation count")
  check_equal(actual.snapshot.source_detail_ceiling, case[3], case[1] .. " ceiling")
  check_equal(actual.snapshot.content_digest_available, case[4], case[1] .. " digest policy")
  check_equal(count_keys(raw.source_refs), 2, case[1] .. " private source retention")

  local rule = record_by_id(actual, "rule:T%C3%B6p")
  local slot = record_by_id(actual, "regex:rule:T%C3%B6p:0")
  check_equal(rule.name, "Töp", case[1] .. " Unicode rule name")
  check_equal(rule.source.span.start_byte, 0, case[1] .. " rule start byte")
  check_equal(rule.source.span.end_byte, 6, case[1] .. " rule end byte")
  check_equal(rule.source.excerpt, "Töp::", case[1] .. " rule excerpt")
  check_equal(slot.facts.pattern, "é", case[1] .. " Unicode regex pattern")
  check_equal(slot.source.span.start_column, 2, case[1] .. " slot start column")
  check_equal(slot.source.span.end_column, 5, case[1] .. " slot end column")
end

local runtime_index = linkedspec.semantic_index(
  read_file(fixture_root .. "runtime.spec"),
  options("runtime.spec", "text")
)
local runtime = materialize_sources(
  semantic_index_module._static_projection_for_testing(runtime_index)
)
local runtime_expected = materialize_sources(runtime_static_expected(model))
check_equal(json.encode(runtime), json.encode(runtime_expected), "runtime-static deep equality")
check_equal(#runtime.records, 7, "runtime-static record count")
check_equal(#runtime.relations, 8, "runtime-static relation count")
check_equal(runtime.snapshot.has_execution, false, "runtime-static execution absence")
check_equal(#records_of_kind(runtime, "execution"), 0, "runtime-static execution records absent")
check_equal(#records_of_kind(runtime, "event"), 0, "runtime-static event records absent")
check_equal(count_relations(runtime, "observed_as"), 0, "runtime-static observations absent")
check_equal(count_relations(runtime, "dispatches_to"), 0, "runtime-static self dispatch absent")
check_equal(count_relations(runtime, "selects_regex"), 2, "runtime-static slot selections")
local runtime_rule = record_by_id(runtime, "rule:Top")
check_equal(runtime_rule.facts.is_repetition, true, "runtime repetition")
check_equal(runtime_rule.facts.rep_min, 2, "runtime repetition minimum")
check_equal(runtime_rule.facts.rep_max, 2, "runtime repetition maximum")
check_equal(record_by_id(runtime, "regex:rule:Top:0").facts.pattern, "a", "runtime first slot")
check_equal(record_by_id(runtime, "regex:rule:Top:1").facts.pattern, "b", "runtime second slot")

local failed_index = linkedspec.semantic_index(
  read_file(fixture_root .. "failed.spec"),
  options("failed.spec", "span")
)
local native_failure = failed_index:compilation_diagnostic():to_json()
check_equal(native_failure.code, "bare_edge_target_undefined", "native failure code preserved")
check_equal(native_failure.stage, "normalize_edges", "native failure stage preserved")
check_equal(
  native_failure.message,
  "bare edge in rule 'Top' targets undefined rule 'Missing'",
  "native failure message preserved"
)
check_equal(native_failure.fields.rule_label, "Top", "native failure rule field")
check_equal(native_failure.fields.target, "Missing", "native failure target field")

local failed_raw = semantic_index_module._static_projection_for_testing(failed_index)
local failed = materialize_sources(failed_raw)
local failed_expected = materialize_sources(expected_snapshot(model, "failed"))
check_equal(json.encode(failed), json.encode(failed_expected), "failed projection deep equality")
check_equal(#failed.records, 6, "failed record count")
check_equal(#failed.relations, 4, "failed relation count")
check_equal(failed.snapshot.state, "failed_compilation", "failed snapshot state")
check_equal(failed.snapshot.source_detail_ceiling, "span", "failed snapshot ceiling")
check_equal(failed.snapshot.content_digest_available, false, "failed snapshot digest policy")
local diagnostic = record_by_id(failed, "diagnostic:compile:0")
check_equal(diagnostic.facts.code, "unknown_rule_reference", "portable failure code")
check_equal(diagnostic.facts.stage, "compile", "portable failure stage")
check_equal(diagnostic.facts.fields.rule_id, "rule:Top", "portable failure rule id")
check_equal(diagnostic.facts.fields.missing_rule_id, "rule:Missing", "portable missing rule id")
check_equal(diagnostic.source.excerpt, "Missing", "portable failure excerpt")
check_equal(diagnostic.source.span.start_byte, 6, "portable failure start byte")
check_equal(diagnostic.source.span.end_byte, 13, "portable failure end byte")
check_equal(record_by_id(failed, "rule:Top").facts.edge_ownership, "action", "failed rule edge")
check_equal(count_relations(failed, "diagnoses"), 1, "failed diagnostic relation")
check_equal(count_relations(failed, "explained_by"), 1, "failed explanation relation")

local repeated_index = linkedspec.semantic_index(
  "Top::\n /x/\n E { return(\"first\") }\n E { return([\"second\"]) }\n",
  options("repeated-lifecycle.spec", "text")
)
local repeated = materialize_sources(
  semantic_index_module._static_projection_for_testing(repeated_index)
)
local lifecycles = records_of_kind(repeated, "lifecycle")
check_equal(#lifecycles, 2, "repeated lifecycle count")
check_equal(lifecycles[1].id, "lifecycle:rule:Top:E:0", "first lifecycle id")
check_equal(lifecycles[2].id, "lifecycle:rule:Top:E:1", "second lifecycle id")
check_equal(lifecycles[1].order, 0, "first lifecycle order")
check_equal(lifecycles[2].order, 1, "second lifecycle order")
check_equal(lifecycles[1].facts.value_shape.kind, "string", "first lifecycle shape")
check_equal(lifecycles[2].facts.value_shape.kind, "array", "second lifecycle shape")
check_equal(lifecycles[1].source.excerpt, 'E { return("first") }', "first lifecycle source")
check_equal(lifecycles[2].source.excerpt, 'E { return(["second"]) }', "second lifecycle source")
check_equal(lifecycles[1].source.span.start_line, 3, "first lifecycle line")
check_equal(lifecycles[2].source.span.start_line, 4, "second lifecycle line")

for _, index in ipairs({
  linkedspec.semantic_index(read_file(fixture_root .. "privacy.spec"), options("privacy.spec", "text")),
  linkedspec.semantic_index(read_file(fixture_root .. "privacy.spec"), options("privacy.spec", "identity")),
  failed_index,
  runtime_index,
}) do
  local projection = semantic_index_module._static_projection_for_testing(index)
  check(plain_and_host_free(projection), "projection contains only plain host-free values")
  check_equal(json.encode(detached(projection)), json.encode(projection), "projection JSON round-trip")
  local encoded = json.encode(projection)
  check(encoded:find("/" .. "Users/", 1, true) == nil, "projection excludes developer-home path")
  check(encoded:find("/" .. "home/", 1, true) == nil, "projection excludes Unix-home path")
  check(encoded:find("/" .. "tmp/", 1, true) == nil, "projection excludes temporary path")
end

failed_raw.records[4].facts.fields.missing_rule_id = "rule:Injected"
local failed_again = semantic_index_module._static_projection_for_testing(failed_index)
check_equal(
  record_by_id(failed_again, "diagnostic:compile:0").facts.fields.missing_rule_id,
  "rule:Missing",
  "failed projection clone detached"
)
check(failed_raw ~= failed_again, "projection materializations are fresh roots")

local parse_failure = linkedspec.semantic_index(
  "Top:::\n /x/\n",
  options("parse-failure.spec", "identity")
)
local parse_projection = semantic_index_module._static_projection_for_testing(parse_failure)
check_equal(#parse_projection.records, 3, "parse-failure record count")
check_equal(#parse_projection.relations, 2, "parse-failure relation count")
check_equal(
  record_by_id(parse_projection, "diagnostic:compile:0").facts.code,
  "semantic_index_parse_failed",
  "parse-failure fallback code"
)

local missing_entry = linkedspec.semantic_index(
  "Top::\n /x/\n",
  options("missing-entry.spec", "identity", "Missing")
)
local missing_projection = semantic_index_module._static_projection_for_testing(missing_entry)
check_equal(#missing_projection.records, 4, "missing-entry record count")
check_equal(#missing_projection.relations, 2, "missing-entry relation count")
check_equal(
  record_by_id(missing_projection, "diagnostic:compile:0").facts.code,
  "entry_rule_not_found",
  "missing-entry fallback code"
)

check_equal(next(failed_index), nil, "index raw state remains empty")
check_equal(getmetatable(failed_index), "protected", "index metatable protected")
local immutable_ok = pcall(function() failed_index.injected = true end)
check_equal(immutable_ok, false, "index mutation rejected")
check_equal(linkedspec._static_projection_for_testing, nil, "root exports no test materializer")
check_equal(linkedspec.semantic_static_projection, nil, "root exports no static projection")
check_equal(linkedspec.semantic_records, nil, "root exports no semantic records")
check_equal(linkedspec.semantic_query, nil, "root exports no semantic query")

local implementation = read_file("lua/src/linkedspec/semantic_static_projection.lua")
for _, forbidden in ipairs({
  'require("linkedspec.spec_parser")',
  'require("linkedspec.compiled_spec")',
  'require("linkedspec.source_emitter")',
  'require("linkedspec.interpreter")',
  'require("linkedspec.matching")',
  "runtime_execute(",
  "execute_generated",
  "diagnostic_sink",
  "semantic_observation",
  "os.getenv",
  "io.open",
  "loadfile",
  "dofile",
  "os.clock",
  "math.random",
}) do
  check(implementation:find(forbidden, 1, true) == nil, "implementation excludes " .. forbidden)
end

if #failures > 0 then
  io.stderr:write("semantic static remaining failures:\n")
  for _, failure in ipairs(failures) do io.stderr:write("- " .. failure .. "\n") end
  os.exit(1)
end

print("semantic static remaining assertions: " .. assertions)
